stock_routing_operation/models/stock_move.py

```diff
@@ -0,0 +1,403 @@
+        move_routing_rules = self.env["stock.routing"]._routing_rule_for_moves(self)
+        for move, rule in move_routing_rules.items():
+            if rule:
+                move.routing_rule_id = rule
+        self._apply_routing_rule_pull()
+        self._apply_routing_rule_push()
+
+    def _action_assign(self):
+        if self.env.context.get("exclude_apply_routing_operation"):
+            super()._action_assign()
+        else:
+            # these methods will call _action_assign in a savepoint
+            # and modify the routing if necessary
+            moves = self._split_and_apply_routing()
+            super(StockMove, moves)._action_assign()
+
+    def _split_and_apply_routing(self):
+        """Apply routing rules
+
+        * calls super()._action_assign() (in a savepoint) on moves not yet
+          available to compute the routing rules
+        * split the moves if their move lines have different source or
+          destination locations and need routing
+        * apply the routing rules (pull and push)
+
+        Important: if you inherit this method to skip the routing for some
+        moves, the method has to return the moves in ``self`` so they are
+        assigned.
+        """
+        moves_routing = self._prepare_routing_pull()
+        if not moves_routing:
+            return self
+        # apply the routing
+        moves = self._routing_splits(moves_routing)
+        moves._apply_routing_rule_pull()
+        moves._apply_routing_rule_push()
+        return moves
+
+    def _prepare_routing_pull(self):
+        """Prepare pull routing rules for moves
+
+        When a move has move lines with different routing rules or lines with
+        routing rules and lines without, on the source/dest location, we have
+        to split the moves. This method assigns the moves in a savepoint to
+        compute the routing rules according the move lines.
+
+        If no routing has to be applied, the savepoint is released.
+        If routing must be applied on at least one move, the savepoint is
+        rollbacked and will be called after the routing rules have been applied.
+
+        Return the computed routing rules for the next step, which will be
+        to split the moves.
+        """
+        if not self:
+            return self
+
+        savepoint_name = uuid.uuid1().hex
+        self.env["base"].flush()
+        # pylint: disable=sql-injection
+        self.env.cr.execute(
+            sql.SQL("SAVEPOINT {}").format(sql.Identifier(savepoint_name))
+        )
+        super()._action_assign()
+
+        moves_routing = self._routing_compute_rules()
+
+        if not any(rule for routing in moves_routing.values() for rule in routing):
+            # no routing to apply, so the reservations done by _action_assign
+            # are valid and we can resolve to a normal flow
+            self.env["base"].flush()
+            # pylint: disable=sql-injection
+            self.env.cr.execute(
```

Comment:
It will be assigned later with the _action_assign anyway right? so the moves will be assigned twice (not sure it has an incidence on the data thought, but if we could avoid it, it would be better)

Reply:
In this execution path, there is no routing to apply, so the savepoint is released and we return from `_action_assign()` (see https://github.com/OCA/stock-logistics-warehouse/pull/788/files/f43ea35ac03a1a86b2e9b6549daa8426d5cd68b8#diff-0295a9825355a3ac40a9eac755697f9bR67-R68)

Reply:
Oh, I see what you mean, I missed this.

Reply:
Fixed